### DISCUSSION TITLE
Cascade deleting records with file field takes enormous amount of CPU

### DISCUSSION THREAD

**Discussion by simonlary** [pocketbase/pocketbase#5246] [Q&A] [answered]
Tested with Pocketbase v0.22.17 on Ubuntu and macOS.

1. Create two collections:
    - `foos` with only a single field (not important)
    - `bars` with two fields
        - `foo`: A relation to the `foos` collection with cascade delete enabled
        - `files`: A multiple files fields
2. Create a couple of `foo` records each with a lot of associated `bar` records that each have multiple files
3. Delete a `foo` record or a couple of `foo` records

The record deletions take a couple of seconds during which a CPU core will probably be 100% used which is expected. But afterward, Pocketbase will start to use every CPU core at 100% for multiple seconds, minutes, or even hours depending on the speed of the device and the number of records/files. Killing and restarting Pocketbase fixes the CPU usage, but the files end up not being actually deleted.

From my testing, the time this cleanup takes increases with the number of files/records to be deleted, which I expected, but it also increases with the number of total files which makes it sound like Pocketbase is iterating over every stored file for every deletion? No idea if this is the case, but it is a weird behaviour.

I can consistently reproduce the issue using [this schema](https://github.com/user-attachments/files/16350283/pb_schema.json), running this JavaScript hook to generate some data, and then deleting a couple of the `foo` records created.

```js
routerAdd("GET", "/generate-data", (c) => {
	const NUMBER_OF_FOOS = 10;
	const NUMBER_OF_BARS_PER_FOO = 300;
	const NUMBER_OF_FILES_PER_BAR = 20;

	const foosCollection = $app.dao().findCollectionByNameOrId("foos");
	const barsCollection = $app.dao().findCollectionByNameOrId("bars");

	for (let i = 0; i < NUMBER_OF_FOOS; i++) {
		const fooRecord = new Record(foosCollection, {
			name: `Foo ${i}`,
		});
		$app.dao().saveRecord(fooRecord);

		for (let j = 0; j < NUMBER_OF_BARS_PER_FOO; j++) {
			const barRecord = new Record(barsCollection);
			const barUpsertForm = new RecordUpsertForm($app, barRecord);
			barUpsertForm.loadData({
				foo: fooRecord.id,
			});
			const files = Array(NUMBER_OF_FILES_PER_BAR)
				.fill(0)
				.map((_, k) => $filesystem.fileFromBytes(Math.random().toString(36), `my-file-${k}.txt`));
			barUpsertForm.addFiles("files", ...files);
			barUpsertForm.submit();
		}
	}

	return c.string(200, "Successfully generated data");
});
```

**Comment by ganigeorgiev** [comment]
Are you sure that the files are the issue and not something else?

Could you confirm that you don't have any custom hooks or 3rd party watch services?

I'll try to setup your example a little bit later but in general files are deleted in a separate goroutine after the records deletion so they shouldn't have affect on the cascade delete operation (at least not on its execution time).

**Reply by ganigeorgiev** [reply]
I've tried to reproduce the issue locally and I'm not able to, even when deleting 5 `foos` records concurrently (aka. ~6000 files). The files deletion is almost instantaneous for me and the CPU usage grows up to 100% just for a second or two and then go back to normal (I'm using a 10+ years old laptop with CPU: `Intel(R) Core(TM) i5-2430M CPU @ 2.40GHz`).


> it sound like Pocketbase is iterating over every stored file for every deletion? No idea if this is the case, but it is a weird behaviour.

This is exactly what is happening. In case of a collection or record deletion we delete all files under the specific prefix (`/:collectionId/*` when deleting collections and `/:collectionId/:recordId/*` when deleting records). Why this is weird? How do you propose doing it?

You can find the code responsible for triggering the files delete at:
https://github.com/pocketbase/pocketbase/blob/966341117193f8060fd85c5d57af92651ce3b1aa/core/base.go#L1145-L1180
and the actual deletion in:
https://github.com/pocketbase/pocketbase/blob/966341117193f8060fd85c5d57af92651ce3b1aa/tools/filesystem/filesystem.go#L278

I agree that the current approach may not be very efficient when using S3 as storage since we'll send individual requests over the network for the removal of each object. I guess one possible optimization for this case could be to postpone the file delete operations and store somewhere information what files need to be deleted and then schedule a job one or twice a day to send a single S3 batch delete request but this approach could introduce fair amount of complexity to the already messy filesystem abstraction and will require extra conditional handling in case the S3 vendor doesn't support or has permissions to execute the `Multi-Object Delete` API, so for now I'm reluctant doing it but other suggestions are welcomed.

**Reply by simonlary** [reply]
> Are you sure that the files are the issue and not something else? Could you confirm that you don't have any custom hooks or 3rd party watch services?

I can reproduce even without any custom hooks or anything. Simply downloading the Pocketbase executable, generate the data using the hook I sent earlier, and then deleting a couple of `foos` records. Also, deleting the `storage` folder containing the files before deleting the records causes the problem to not happen.

> in general files are deleted in a separate goroutine after the records deletion so they shouldn't have affect on the cascade delete operation

Yeah that is working perfectly, the actual record deletion is very fast for the number of records. My issue is the goroutine that runs afterward often takes the whole computer down. 😕 

> This is exactly what is happening. In case of a collection or record deletion we delete all files under the specific prefix (/:collectionId/* when deleting collections and /:collectionId/:recordId/* when deleting records). Why this is weird? How do you propose doing it?

I meant that it seems like deleting the files for a record goes over all the files in `/:collectionId/*`, not simply in `/:collectionId/:recordId/*`. Going over all the files under `/:collectionId/:recordId/*` is what I expected, but it looks like the time to delete the files scales proportionnaly to the _total_ number of files in all the records in the collection, which is the part I find weird. 🙂 

> I agree that the current approach may not be very efficient when using S3 as storage since we'll send individual requests over the network for the removal of each object.

I'm only storing files locally, so network requests are not the problem.

**Reply by ganigeorgiev** [reply]
Another optimization that wouldn't complicate too much the existing filesystem implementation I guess could be instead of firing immediately the delete goroutine for each prefix to store somewhere in the database what records were deleted (aka. prefixes to delete) and then iterate over them on batches per 100 or something similar on every couple minutes to prevent the CPU spikes. This could also prevent hitting the S3 API rate-limits/throttling and could allow supporting fail-retry of the delete operation (currently in the rare case the file delete fails we just log it since it happens in the background).

I've added it in my todo and will consider it either as part or after the ongoing refactoring.

edit: it seems that we commented near the same time but again as mentioned earlier locally I'm not able to reproduce it and the CPU is definitely not spiking for _"minutes or hours"_ to perform the files deletion. Make sure that there is no 3rd party service or dir watcher that could be affected by the multiple file delete events.

**Reply by simonlary** [reply]
> locally I'm not able to reproduce it and the CPU is definitely not spiking for "minutes or hours" to perform the files deletion.

The sample I created doesn't take minutes or hours, but it still takes multiple seconds using all 12 cores at 100% on an M2 MacBook Pro which is a ton of CPU usage for deleting about 300 directories/6000 files. The "minutes or hours" is in my production case where I have the equivalent of 5000 `foos`, 200 000 `bars`, and 150 000 files. I would assume deleting 6000 files in that cases to be as long as deleting 6000 files in my example since there is about the same number of files per `bar`, but it takes waay longer. Which is why I assume that something somewhere is scaling the file deletion time depending on the total number of files in the collection. Simply increasing the `NUMBER_OF_FOOS` in the data generation will make the deletion take longer.

> Make sure that there is no 3rd party service or dir watcher that could be affected by the multiple file delete events.

The pocketbase process is the one that takes the whole CPU usage.

I'll try to look more into it, but is it possible that maybe a new thread is created for every file to delete? Or every time a record is deleted, pocketbase lists all the files in `/:collectionId/*` or something similar?

**Reply by ganigeorgiev** [reply]
> I'll try to look more into it, but is it possible that maybe a new thread is created for every file to delete? 

It is up to the Go runtime to determine where to run the fired goroutine. We don't try to manage it or have control over it.

> Or every time a record is deleted, pocketbase lists all the files in /:collectionId/* or something similar?

On closer look on the [gocloud.dev/fileblob](https://github.com/google/go-cloud/blob/fc42ff384c071261d064e99e2b95295dc6ab6338/blob/fileblob/fileblob.go#L418-L421) implementation you may be right. In the above code the check seems to expect the prefix to end with `/` otherwise it will start walking from the directory before the last `/` slash (in our case of `:collectionId/:recordId` this would mean `:collectionId` for every deleted record).

I'll test it a little bit later and eventually will add a trailing slash to the prefix to see if it will have any effect.

**Reply by ganigeorgiev** [reply]
Yes, I could confirm that adding a trailing slash to the `DeletePrefix` argument reduces significantly the CPU usage when trying to delete 30 `foos` records at once (~180000 files).

I'll submit a minor release sometime later today or tomorrow in the morning.

**Reply by simonlary** [reply]
> Yes, I could confirm that adding a trailing slash to the `DeletePrefix` argument reduces significantly the CPU usage when trying to delete 30 `foos` records at once (~180000 files).
> 
> I'll submit a minor release sometime later today or tomorrow in the morning.

Awesome! I'll wait for that release then. 😁

**Reply by ganigeorgiev** [reply, marked as answer]
The above fix for the local filesystem is available in [v0.22.18 release](https://github.com/pocketbase/pocketbase/releases/tag/v0.22.18).
As mentioned in https://github.com/pocketbase/pocketbase/discussions/5246#discussioncomment-10128428, a more general optimization that should help when S3 storage is used will be also implemented in the future.

**Reply by simonlary** [reply]
Just tested it last night on our real data and instead of bringing our VM down after deleting 5 or 10 records it managed to delete all 3000 records it had to without any problem! 🎉 

Thank you very much for the help and the insanely fast responses!
